lib/spack/spack/environment.py

```diff
@@ -854,9 +944,16 @@ def write(self):
         # invalidate _repo cache
         self._repo = None
 
+        # put any changes in the definitions in the YAML
```

Comment:
I have a question about this, should the changes outlined in the definitions be written to disk? Would this make it harder to share the stack if one leverages `when`.

Reply:
@paulbry
The definitions are effectively building blocks for the list of abstract specs. So they get written to disk the same way that the abstract specs do.

The use case is something like "I want to add `gcc@8.1.0` to the list of compilers I use.

`spack add -l compilers %gcc@8.1.0`

Now every package that is in a matrix against the `$compilers` list gets built with `gcc@8.1.0` in addition to all of the other compilers.

Reply:
@becker33 
Thanks for the details, makes sense.

The main reason I was wondering is that when I run something like `spack env create name envfile` if the `envfile` used had the following:

```
  definitions:
  - compilers: ['%gcc@8.3.0', '%clang@3.4.2']
  - when: arch == 'linux-centos7-ppc64le'
    compilers: ['%gcc@4.8.5']
```

Then the resulting `spack.yaml` file would look like this:

```
  definitions:
  - compilers: ['%gcc@8.3.0', '%clang@3.4.2', '%gcc@4.8.5']
  - when: arch == 'linux-centos7-ppc64le'
    compilers: ['%gcc@4.8.5']
```

The same occurs if I manually edited the `spack.yaml` definitions file within an activate environment and performed some operation. This just seems to present a difficult workflow in terms of sharing an environment with someone else who could potentially have different definitions when `when` is involved.

I'm not really passionate about how that workflow, just something I had noticed when using it last week. If I develop with and share the `envfile` there isn't that concern.

Reply:
@paulbry

Thanks for pointing that out!

I've updated the PR to handle this properly.

New specs will be added to the first entry of that name for which the `when` clause, if present, is satisfied. Specs should be properly removed from whichever list they occurred in.

The new commit includes a test for this behavior, but can you please confirm that it works for you as well?

Reply:
@becker33 
Thanks, looks like that fix works as expected.